provider/oci/environ.go

```diff
@@ -223,42 +378,313 @@ func (e *Environ) Destroy(ctx envcontext.ProviderCallContext) error {
+		},
+	)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	tools, err := args.Tools.Match(tools.Filter{Arch: spec.Image.Arch})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	logger.Tracef("agent binaries: %v", tools)
+	if err = args.InstanceConfig.SetTools(tools); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	if err = instancecfg.FinishInstanceConfig(
+		args.InstanceConfig,
+		e.Config(),
+	); err != nil {
+		return nil, errors.Trace(err)
+	}
+	hostname, err := e.namespace.Hostname(args.InstanceConfig.MachineId)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	tags := args.InstanceConfig.Tags
+
+	var apiPort int
+	var desiredStatus ociCore.InstanceLifecycleStateEnum
+	// Wait for controller to actually be running
+	if args.InstanceConfig.Controller != nil {
+		apiPort = args.InstanceConfig.Controller.Config.APIPort()
+		desiredStatus = ociCore.InstanceLifecycleStateRunning
+	} else {
+		desiredStatus = ociCore.InstanceLifecycleStateProvisioning
+	}
+
+	cloudcfg, err := e.getCloudInitConfig(series, apiPort)
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot create cloudinit template")
+	}
+
+	// compose userdata with the cloud config template
+	logger.Debugf("Composing userdata")
+	userData, err := providerinit.ComposeUserData(
+		args.InstanceConfig,
+		cloudcfg,
+		OCIRenderer{},
+	)
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot make user data")
+	}
+
+	var rootDiskSize int
+	if args.Constraints.RootDisk != nil {
+		rootDiskSize = int(*args.Constraints.RootDisk)
+		if int(*args.Constraints.RootDisk) < MinVolumeSizeMB {
+			rootDiskSize = MinVolumeSizeMB / 1024
+		} else if int(*args.Constraints.RootDisk) > MinVolumeSizeMB {
+			rootDiskSize = MaxVolumeSizeMB / 1024
+		}
+	} else {
+		rootDiskSize = MinVolumeSizeMB / 1024
+	}
+
+	// NOTE(gsamfira): Should we make this configurable?
+	// TODO(gsamfira): select Availability Domain and subnet ID
+	assignPublicIp := true
+	bootSource := ociCore.InstanceSourceViaImageDetails{
+		ImageId:             &image,
+		BootVolumeSizeInGBs: &rootDiskSize,
+	}
+	instanceDetails := ociCore.LaunchInstanceDetails{
+		AvailabilityDomain: &zone,
+		CompartmentId:      e.ecfg().compartmentID(),
+		SourceDetails:      bootSource,
+		Shape:              &spec.InstanceType.Name,
+		CreateVnicDetails: &ociCore.CreateVnicDetails{
+			SubnetId:       network.Id,
+			AssignPublicIp: &assignPublicIp,
+			DisplayName:    &hostname,
+		},
+		DisplayName: &hostname,
+		Metadata: map[string]string{
+			"user_data": string(userData),
+		},
+		FreeformTags: tags,
+	}
+
+	request := ociCore.LaunchInstanceRequest{
+		LaunchInstanceDetails: instanceDetails,
+	}
+
+	response, err := e.Compute.LaunchInstance(context.Background(), request)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	instance, err := newInstance(response.Instance, e)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	machineId := response.Instance.Id
+	timeout := 10 * time.Minute
```

Comment:
we have other config around how long to wait for a machine to be accessible.
./bootstrap/config.go:36:       BootstrapTimeoutKey = "bootstrap-timeout"

Can we just use something like that here?
I didn't think we timed out for launching any other instances other than the bootstrap machine.

Reply:
ping for using BootstrapTimeout here.

Reply:
Looked into it. Can't really use this. It's not really made available to providers, and only seems to be used here:

https://github.com/juju/juju/blob/develop/cmd/juju/commands/bootstrap.go#L640